zcash_client_backend/src/scan.rs

```diff
@@ -123,30 +128,35 @@ impl<D: BatchDomain, Output: ShieldedOutput<D, COMPACT_NOTE_SIZE> + Clone> Batch
 type ResultKey = (BlockHash, TxId);
 
 /// Logic to run batches of trial decryptions on the global threadpool.
-pub(crate) struct BatchRunner<D: BatchDomain, Output: ShieldedOutput<D, COMPACT_NOTE_SIZE>> {
+pub(crate) struct BatchRunner<A, D: BatchDomain, Output: ShieldedOutput<D, COMPACT_NOTE_SIZE>> {
     batch_size_threshold: usize,
-    acc: Batch<D, Output>,
-    pending_results: HashMap<ResultKey, channel::Receiver<OutputIndex<Option<DecryptedNote<D>>>>>,
+    acc: Batch<A, D, Output>,
+    pending_results: HashMap<ResultKey, OutputResult<A, D>>,
 }
 
-impl<D, Output> BatchRunner<D, Output>
+impl<A, D, Output> BatchRunner<A, D, Output>
 where
+    A: Clone,
     D: BatchDomain,
     Output: ShieldedOutput<D, COMPACT_NOTE_SIZE>,
-    D::IncomingViewingKey: Clone,
 {
     /// Constructs a new batch runner for the given incoming viewing keys.
-    pub(crate) fn new(batch_size_threshold: usize, ivks: Vec<D::IncomingViewingKey>) -> Self {
+    pub(crate) fn new(
+        batch_size_threshold: usize,
+        ivks: impl Iterator<Item = (A, D::IncomingViewingKey)>,
+    ) -> Self {
+        let (tags, ivks) = ivks.unzip();
         Self {
             batch_size_threshold,
-            acc: Batch::new(ivks),
+            acc: Batch::new(tags, ivks),
             pending_results: HashMap::default(),
         }
     }
 }
 
-impl<D, Output> BatchRunner<D, Output>
+impl<A, D, Output> BatchRunner<A, D, Output>
 where
+    A: Clone + Send + 'static,
```

Comment:
Why `'static`?

Reply:
This is necessary to use it with `rayon`:
![image](https://user-images.githubusercontent.com/4993799/190019886-5ff917b0-1fdd-4ec3-b04a-574f60510806.png)

Reply:
Specifically, it's because we spawn a new task, which may outlive the current thread. Thus we need to ensure that `A` contains no lifetimes that could outlive the current thread.

Reply:
👍